### PR TITLE
[3.10] Make sure the update hash checker produces the correct result

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

- Only show a message when the verification fails (request by @mbabker )
- Handle the verification locally in com_joomlaupdate so we don't depend on the other method to be up-to-date. (like releasing a standalone version)
- Throw something in the log when the verification fails.
- Make sure we are using the correct update object and don't do some magic with passing the update xml.

###### First Case 
- install 3.9-dev https://developer.joomla.org/nightlies/Joomla_3.9.0-dev-Development-Full_Package.zip
- Apply this update server: https://www.jah-tz.de/downloads/core/nightlies/next_minor_list.xml
- install the update
- notice there is a warning claiming that the hash was not set. ;)

###### Seccond Case
- Install 3.9-dev with this changes: https://github.com/zero-24/joomla-cms/archive/fix39updater.zip
- Apply this update server: https://www.jah-tz.de/downloads/core/nightlies/next_minor_list.xml
- install the update
- notice there is no warning

### Expected result

Make sure the update works without any kind of warning

### Actual result

Well the update works but throws a warning as the url is not passed.

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by zero-24** [comment]
Tagging @alikon too ;)

File: administrator/components/com_joomlaupdate/controllers/update.php [hunk 1]
```diff
@@ -50,20 +50,20 @@ public function download()
 		$message     = null;
 		$messageType = null;
 
-		switch ($result['check'])
+		// The validation was not successful for now just a warning for now.
```
Comment (mbabker) [inline administrator/components/com_joomlaupdate/controllers/update.php]: Same thing as the other PR, leave a comment here for 4.0 to put the switch case back.

File: administrator/components/com_joomlaupdate/models/default.php [hunk 2]
```diff
@@ -331,10 +330,44 @@ public function download()
...
 		}
 
-		$response['check'] = JInstallerHelper::isChecksumValid($target, $updateInfo['url']);
+		$response['check'] = $this->isChecksumValid($target, $updateInfo['object']);
+
 		return $response;
 	}
 
+	/**
+	 * Return the result of the checksum of a package with the SHA256/SHA384/SHA512 tags in the update server manifest
```
Comment (wilsonge) [inline administrator/components/com_joomlaupdate/models/default.php:339]: Just add a note why we have 'forked' this method here (given it's a duplicate by eye)

**Comment by alikon** [comment]
> Handle the verification locally in com_joomlaupdate so we don't depend on the other method to be up-to-date. (like releasing a standalone version)

Don't understand (but it's me ;) )  , can you give me a pratical example?

**Comment by mbabker** [comment]
We are able to release an update of `com_joomlaupdate` standalone from Joomla core in case there are cases where the update component needs to be updated before running a core update (this has happened before).  So, the update component can't rely 100% on the core Installer or Updater libraries, especially as new features (such as this) are introduced, inlining (duplicating) this code within the component ensures if we do have a situation in the future where a standalone package has to be built that if it's applied to a Joomla version where this feature (or a similar case for future feature additions) doesn't exist it doesn't cause fatal errors.

**Comment by alikon** [comment]
make sense thanks for the explanation

**Review by alikon — APPROVED** [review]

File: administrator/components/com_joomlaupdate/models/default.php (hunk 2 above)
Comment (zero-24) [reply administrator/components/com_joomlaupdate/models/default.php:339]: Done with the last commit. thanks

File: administrator/components/com_joomlaupdate/controllers/update.php (hunk 1 above)
Comment (zero-24) [reply administrator/components/com_joomlaupdate/controllers/update.php]: Done with the last commit. thanks

**Review by brianteeman — APPROVED** [review]

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 4fd66f6d63cca9a03052ad461ed60e5fd63a08f6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20329">issues.joomla.org/tracker/joomla-cms/20329</a>.</sub>

**Comment by ReLater** [comment]
>  Install 3.9-dev with this changes: https://github.com/zero-24/joomla-cms/archive/fix39updater.zip

When I install this package com_joomlaupdate says "You already have the latest Joomla version, **3.10.0**-dev". Shouldn't it be **3.9-dev** for testing?

Or is this pr already RTC?

**Comment by Quy** [comment]
@ReLater PR is not RTC yet.

**Comment by ReLater** [comment]
@Quy Thank you!
@zero-24 Could you take a look, please?

**Comment by brianteeman** [comment]
(updated title to show it is now for 3.10)

**Comment by mbabker** [comment]
It's actually intended for 3.9 but because of [the branch changes](https://developer.joomla.org/news/728-updated-branches-on-joomla-repository.html) it had to be moved to the 3.10-dev branch otherwise the PR would've been totally broken.  Test this as is, when RTC the same patch will apply to the 3.9-dev branch.

**Comment by ReLater** [comment]
> Test this as is

Maybe i don't understand correctly. Step "Second Case" of testing instructions is not testable for me as is because https://github.com/zero-24/joomla-cms/archive/fix39updater.zip installs a Joomla 3.10 (instead of expected 3.9). Afterwards com_joomlaupdate with update server https://www.jah-tz.de/downloads/core/nightlies/next_minor_list.xml doesn't offer the possibility to reinstall an update package because only 3.9 versions are included in xml file.

**Comment by zero-24** [comment]
Moved to: https://github.com/joomla/joomla-cms/pull/20644

**Pull request closed without merging**